### PR TITLE
add restart policy & scheduler name for workflow pods

### PR DESCRIPTION
**Pull request by houz42**
RestartPolicy and SchedulerName are useful for controlling pods running of workflow.

### CODE REVIEW

File: pkg/apis/workflow/v1alpha1/types.go [hunk 1]
```diff
@@ -135,6 +135,18 @@ type WorkflowSpec struct {
 	ActiveDeadlineSeconds *int64 `json:"activeDeadlineSeconds,omitempty"`
 	// Priority is used if controller is configured to process limited number of workflows in parallel. Workflows with higher priority are processed first.
 	Priority *int32 `json:"priority,omitempty"`
+
+	// Restart policy for all containers within all the pods.
+	// One of Always, OnFailure, Never.
```
Comment (alexmt) [inline pkg/apis/workflow/v1alpha1/types.go]: By default workflow controller creates pods with restart policy `Never`

File: pkg/apis/workflow/v1alpha1/types.go [hunk 2]
```diff
@@ -135,6 +135,18 @@ type WorkflowSpec struct {
...
+	// Default to Always.
+	// More info: https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#restart-policy
+	// +optional
+	RestartPolicy apiv1.RestartPolicy `json:"restartPolicy,omitempty"`
```
Comment (alexmt) [inline pkg/apis/workflow/v1alpha1/types.go]: I think we should be consistent with Job and allow only `Never` or `OnFailure` values.

**Comment by houz42** [comment]
@alexmt 
1. scheduler name and restart policy have been made step level
2. restart policy has been restricted to Never and OnFailure

**Comment by jessesuen** [comment]
@houz42 I think the scheduler is a fine addition.

However a restartPolicy of OnFailure is problematic to set because restartPolicy is a pod spec level setting, and it will apply to the wait sidecar as well. The current design is that the controller relies on the fact the wait sidecar will exit with non-zero in many situations to understand the status of the step. For example, the entire wait logic will non-zero if any of the following goes wrong:

* artifact loading
* log retrieval
* output parameter retrieval
* artifact saving
* output annotation
* wait k8s resource reached failure condition

In order to support a restartPolicy of OnFailure, we would need to modify the executor to *always* exit zero, and communicate back to the controller, that a step had failed in a different way. It's unclear what this mechanism would be.

One thought is: we currently already use pod annotations to communicate error *messages* to the controller. The controller could be modified such that it *always* expects a pod annotation to be set (even on success). Then, if the pod completed without setting an annotation, then something went wrong and the controller could fail the step.

So as it stands, supporting `restartPolicy: OnFailure` can't go in without first replacing the current `exit 1` error communication mechanism with something else.

**Comment by houz42** [comment]
@jessesuen maybe I should submit changes on schedulerName first and consider on restartPolicy later.

**Comment by jessesuen** [comment]
Yes, scheduler only changes would be fine.

**Comment by houz42** [comment]
add scheduler name only in #1184

**Pull request closed without merging**